src/Substrait.Core/ProtoConverter.cs

```diff
@@ -0,0 +1,154 @@
+﻿using Substrait.Expression;
+using Substrait.Protobuf;
+using Substrait.Relation;
+using System.Collections.Immutable;
+using ProtoRel = Substrait.Protobuf.Rel;
+using ProtoRex = Substrait.Protobuf.Expression;
+using Rel = Substrait.Relation.Rel;
+
+namespace Substrait.Core
+{
+  /// <summary>
+  /// Utility to convert Substrait protobuf messages to/from charp domain-objects.
+  /// </summary>
+  public class ProtoConverter
```

Comment:
One possible recommendation for the `ProtoConverter` to split work in multiple small PRs while at the same time making committed code self-contained is to divide the work across operator boundaries, e.g., take scan operator and translate all its existing contents from proto into internal representation, including unit tests to validate the translation using a few examples (maybe consisting of a single op).

Said that, it might still be fine to push this PR since we are in very early stages of the library development (I'd like to know what @westonpace thinks too) but maybe add some comments in the files with a TODO/warning note so there is a clear understanding that translation has some significant gaps (as it is already pointed out in the PR title)? Thoughts?

Reply:
I think a comment, or even throwing a `NotImplementedException` is probably fine.  But I think whatever feels natural is fine.  This seems like a rather arbitrary swath of functionality in this PR but, on the other hand, it is enough to start sparking some initial feedback.

I've found that is quite quick to get to 80% coverage and then the last 20% ends up being quite tricky.  For example, user defined types and extension relations will probably require more thought.  Going operator by operator can be challenging because, for example, a project relation needs an expression which could be anything (and, as mentioned, user defined type expressions might come quite a bit later).

Reply:
@jcamachor, I see value in your suggestion, particularly from tracking and reviews. Anyone who checks out this commit would know that what features will work. I started it like that with just `inputs` to start with. But I can pay more attention going forward.